File: app/src/main/java/com/johnpetitto/orachat/TimeUtils.java
package com.johnpetitto.orachat;

import android.content.Context;
import android.support.annotation.PluralsRes;
import android.util.Log;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.concurrent.TimeUnit;

public final class TimeUtils {
    private TimeUtils() {}

    public static final DateFormat dateFormatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssXX", Locale.ENGLISH);

    private static final long SECONDS = TimeUnit.SECONDS.toMillis(1);
    private static final long MINUTES = TimeUnit.MINUTES.toMillis(1);
    private static final long HOURS = TimeUnit.HOURS.toMillis(1);
    private static final long DAYS = TimeUnit.DAYS.toMillis(1);
    private static final long YEARS = TimeUnit.DAYS.toMillis(365);

    public static String getTimeAgo(Context context, String createdAt) {
        try {
            long currentTime = new Date().getTime();
            long createdAtTime = dateFormatter.parse(createdAt).getTime();
            long diff = currentTime - createdAtTime;

            if (diff < MINUTES) {
                return getTimeSuffix(context, R.plurals.seconds_suffix, diff / SECONDS);
            } else if (diff < HOURS) {
                return getTimeSuffix(context, R.plurals.minutes_suffix, diff / MINUTES);
            } else if (diff < DAYS) {
                return getTimeSuffix(context, R.plurals.hours_suffix, diff / HOURS);
            } else if (diff < YEARS) {
                return getTimeSuffix(context, R.plurals.days_suffix, diff / DAYS);
            } else {
                return getTimeSuffix(context, R.plurals.years_suffix, diff / YEARS);
            }
        } catch (ParseException e) {
            Log.e("TimeUtils", e.getMessage());
        }

        return null;
    }

    private static String getTimeSuffix(Context context, @PluralsRes int resId, long quantity) {
        return context.getResources().getQuantityString(resId, (int) quantity, quantity);
    }
}

File: app/src/main/java/com/johnpetitto/orachat/TimeUtils.kt
package com.johnpetitto.orachat

import android.content.Context
import android.support.annotation.PluralsRes
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeUnit

val dateFormatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssXX", Locale.ENGLISH)

private val secs = TimeUnit.SECONDS.toMillis(1)
private val mins = TimeUnit.MINUTES.toMillis(1)
private val hours = TimeUnit.HOURS.toMillis(1)
private val days = TimeUnit.DAYS.toMillis(1)
private val months = TimeUnit.DAYS.toMillis(30)
private val years = TimeUnit.DAYS.toMillis(365)

fun getTimeAgo(context: Context, createdAt: String): String {
    val currentTime = Date().time
    val createdAtTime = dateFormatter.parse(createdAt).time
    val diff = currentTime - createdAtTime

    return when {
        diff < secs -> context.getString(R.string.now)
        diff < mins -> getTimeSuffix(context, R.plurals.seconds_suffix, diff / secs)
        diff < hours -> getTimeSuffix(context, R.plurals.minutes_suffix, diff / mins)
        diff < days -> getTimeSuffix(context, R.plurals.hours_suffix, diff / hours)
        diff < months -> getTimeSuffix(context, R.plurals.days_suffix, diff / days)
        diff < years -> getTimeSuffix(context, R.plurals.months_suffix, diff / months)
        else -> getTimeSuffix(context, R.plurals.years_suffix, diff / years)
    }
}

private fun getTimeSuffix(context: Context, @PluralsRes resId: Int, quantity: Long) =
        context.resources.getQuantityString(resId, quantity.toInt(), quantity)
